driver-core/src/main/com/mongodb/binding/SingleServerBinding.java

```diff
@@ -111,13 +111,15 @@ public Connection getConnection() {
         @Override
         public ConnectionSource retain() {
             super.retain();
-            SingleServerBinding.this.retain();
             return this;
         }
 
         @Override
         public void release() {
-            SingleServerBinding.this.release();
+            super.release();
+            if (super.getCount() == 0) {
```

Comment:
It made more sense to only retain and release the binding when the source is created and decremented back to 0.

Reply:
Ack